File: khandaq-ios/Antidote/KhandaqPush.swift
// Khandaq 0.2.0 push relay constants
import Foundation
import CommonCrypto

enum KhandaqPush {
    static let relayBase = "https://push.khandaq.org"
    static let fcmPushURLPrefix = relayBase + "/toxfcm/fcm.php?id="

    /// Set via Info.plist KhandaqPushRelayAuthSecret (optional).
    private static let relayAuthSecret: String = {
        Bundle.main.object(forInfoDictionaryKey: "KhandaqPushRelayAuthSecret") as? String ?? ""
    }()

    static func pushURL(forFcmToken token: String) -> String {
        fcmPushURLPrefix + token + "&type=1"
    }

    static func isAllowedPushURL(_ url: String) -> Bool {
        guard let components = URLComponents(string: url),
              let host = components.host?.lowercased() else {
            return false
        }
        let path = components.path

        if host == "push.khandaq.org" {
            return path == "/toxfcm/fcm.php" && (components.queryItems?.first(where: { $0.name == "id" })?.value?.count ?? 0) >= 10
        }
        if host == "tox.zoff.xyz" {
            return path == "/toxfcm/fcm.php" && (components.queryItems?.first(where: { $0.name == "id" })?.value?.count ?? 0) >= 10
        }
        return false
    }

    /// Append sender pubkey and the replay-resistant relay auth (security NEW-2).
    /// Must match the server (infra/push/relay/app.py) and the objcTox push builder
    /// (OCTSubmanagerChatsImpl khandaqAppendRelayAuth) byte-for-byte:
    ///   msg  = id + "\n" + from + "\n" + ts   (raw, URL-decoded values, UTF-8)
    ///   auth = lowercase hex HMAC-SHA256(secret, msg)
    /// NOTE: the live push is actually sent from the objcTox pod (which can't reach this
    /// app-target Swift); this stays in sync as the canonical reference / for any Swift caller.
    static func withWakeParams(_ url: String, senderPubkey: String?) -> String {
        var result = url
        if let pk = senderPubkey, !pk.isEmpty, !result.contains("from=") {
            result += result.contains("?") ? "&from=\(pk)" : "?from=\(pk)"
        }

        guard !relayAuthSecret.isEmpty,
              result.contains("push.khandaq.org"),
              !result.contains("auth=") else {
            return result
        }

        let idValue = queryValue(result, "id")
        let fromValue = queryValue(result, "from")
        let ts = String(Int(Date().timeIntervalSince1970))
        let auth = hmacSha256Hex(relayAuthSecret, "\(idValue)\n\(fromValue)\n\(ts)")
        if !auth.isEmpty {
            result += "&ts=\(ts)&auth=\(auth)"
        }
        return result
    }

    /// URL-decoded value of a query parameter, to match the server's decoded view.
    private static func queryValue(_ url: String, _ key: String) -> String {
        guard let components = URLComponents(string: url) else { return "" }
        return components.queryItems?.first(where: { $0.name == key })?.value ?? ""
    }

    // MARK: - Per-contact capabilities (re-audit 2026-08-22, K-01)
    //
    // The shared push HMAC is baked into every published binary, so unpacking one yields a
    // credential that wakes any device whose FCM token you also hold. The fix is a secret scoped to
    // one relationship: this device mints 32 random bytes per contact, registers their SHA-256 with
    // the relay, and publishes them inside the wake URL that contact already receives over Tox.
    //
    // Registration cannot be authorised by the shared secret — every installation has it — and it
    // cannot be authorised by knowing the FCM token either, because every contact knows that too:
    // it is in the wake URL they were handed. So the relay pushes a nonce TO THE TOKEN, data-only,
    // and only the device that owns that FCM registration can echo it back. A contact can start a
    // challenge and can never finish one.
    //
    // The result is written to UserDefaults because the live push is emitted from the objcTox pod,
    // which cannot reach app-target Swift. `OCTPushUrlValidator.ownWakeURLForToken:friendPublicKey:`
    // reads exactly these keys.

    /// Prefix for one contact's capability. Must match OCTPushUrlValidator.m.
    private static let capPrefix = "khandaq_pushcap_"
    /// The FCM token the stored capabilities were registered against.
    private static let capTokenKey = "khandaq_pushcap_token"

    // MARK: Where capabilities live (re-review 2026-08-22, KQ-08)
    //
    // These are bearer-style per-contact authorization secrets: hold one and you can make the relay
    // wake that device for that relationship until it is revoked or expires. UserDefaults is
    // app-private under normal sandboxing and is still a PREFERENCES store — it lands in backups and
    // in anything that reads the app container. Keychain, device-only, is where a bearer secret
    // belongs.
    //
    // Accessible-after-first-unlock rather than when-unlocked: the push path runs in the background,
    // including right after a reboot before the user has unlocked, and a capability that cannot be
    // read then is a notification that never arrives. ThisDeviceOnly keeps it out of backups and off
    // a restored device, which is the property the review asks for.
    private static let capKeychainService = "org.khandaq.pushcap"

    private static func keychainRead(_ account: String) -> String? {
        var query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: capKeychainService,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        // NOTE: kSecAttrAccessible is deliberately absent here. In a query it acts as a SEARCH
        // predicate, so including it would fail to find items written under any other class — the
        // exact bug that once logged this app out on every launch (see KeychainManager).
        var item: CFTypeRef?
        let status = withUnsafeMutablePointer(to: &item) {
            SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer($0))
        }
        query.removeAll()
        guard status == errSecSuccess, let data = item as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    @discardableResult
    private static func keychainWrite(_ account: String, _ value: String) -> Bool {
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: capKeychainService,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(base as CFDictionary)
        var add = base
        add[kSecValueData as String] = Data(value.utf8)
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        return SecItemAdd(add as CFDictionary, nil) == errSecSuccess
    }

    private static func keychainDelete(_ account: String) {
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: capKeychainService,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(base as CFDictionary)
    }

    /// One-time move of a value written by a build that used UserDefaults.
    ///
    /// Runs on read, not at launch: a launch-time sweep would have to enumerate every contact, and
    /// the only capabilities that matter are the ones something actually asks for. The preference is
    /// removed only AFTER the Keychain write succeeds — the reverse order would lose a capability on
    /// a device where the Keychain is briefly unavailable, and a lost capability is a contact whose
    /// pushes stop until the recipient re-registers.
    private static func migrateFromDefaultsIfNeeded(_ account: String) -> String? {
        let defaults = UserDefaults.standard
        guard let legacy = defaults.string(forKey: account), !legacy.isEmpty else { return nil }
        if keychainWrite(account, legacy) {
            defaults.removeObject(forKey: account)
            return legacy
        }
        // Keychain refused: keep the preference so nothing is lost, and try again next time.
        return legacy
    }

    private static func capabilityValue(_ account: String) -> String? {
        if let fromKeychain = keychainRead(account), !fromKeychain.isEmpty {
            return fromKeychain
        }
        return migrateFromDefaultsIfNeeded(account)
    }

    private static let capQueue = DispatchQueue(label: "org.khandaq.pushcap")
    /// Challenge id -> continuation waiting for its nonce.
    private static var pendingChallenges: [String: (String) -> Void] = [:]
    /// When registration for a contact was last attempted, so a relay outage cannot spin.
    private static var lastAttempt: [String: Date] = [:]
    private static let retryCooldown: TimeInterval = 15 * 60

    static func capabilityKey(forFriend publicKey: String) -> String {
        capPrefix + publicKey.uppercased()
    }

    /// The capability already issued to one contact, or nil.
    ///
    /// nil means "publish the URL without one", which is the pre-capability behaviour and is always
    /// safe: the relay never requires a capability it was not given.
    static func capability(forFriend publicKey: String, ownToken: String) -> String? {
        guard let registeredFor = capabilityValue(capTokenKey), registeredFor == ownToken else {
            // The FCM token rotated. Every capability was registered against the old one and can
            // never be used again; publishing one would hand the relay something it does not know.
            return nil
        }
        let cap = capabilityValue(capabilityKey(forFriend: publicKey))
        return (cap?.isEmpty == false) ? cap : nil
    }

    /// Mint, register and store a capability for one contact. Calls back on an arbitrary queue.
    ///
    /// Failure is not an error condition for the caller: the URL is then published without a
    /// capability, exactly as before this existed.
    static func issueCapability(forFriend publicKey: String, ownToken: String,
                                completion: @escaping (String?) -> Void) {
        if let existing = capability(forFriend: publicKey, ownToken: ownToken) {
            completion(existing)
            return
        }
        var proceed = false
        capQueue.sync {
            if let last = lastAttempt[publicKey], Date().timeIntervalSince(last) < retryCooldown {
                proceed = false
            } else {
                lastAttempt[publicKey] = Date()
                proceed = true
            }
        }
        guard proceed else {
            completion(nil)
            return
        }

        var raw = [UInt8](repeating: 0, count: 32)
        guard SecRandomCopyBytes(kSecRandomDefault, raw.count, &raw) == errSecSuccess else {
            completion(nil)
            return
        }
        // base64url, unpadded: '@' and '=' anywhere in a wake URL make it unroutable through the
        // shipped validators (host-confusion defence A33), so the encoding is load-bearing.
        let cap = Data(raw).base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")

        postJSON(relayBase + "/register/challenge", ["token": ownToken]) { response in
            guard let cid = response?["cid"] as? String, !cid.isEmpty else {
                completion(nil)
                return
            }
            // Wait for the relay's data push. Bounded: a device with notifications switched off at
            // the OS level will never receive one, and that must degrade to "no capability", not to
            // a callback that never fires.
            var settled = false
            let finish: (String?) -> Void = { nonce in
                guard !settled else { return }
                settled = true
                capQueue.sync { pendingChallenges[cid] = nil }
                guard let nonce = nonce else {
                    completion(nil)
                    return
                }
                postJSON(relayBase + "/register/confirm",
                         ["cid": cid, "nonce": nonce, "cap": cap]) { confirmed in
                    guard confirmed != nil else {
                        completion(nil)
                        return
                    }
                    guard keychainWrite(capabilityKey(forFriend: publicKey), cap),
                          keychainWrite(capTokenKey, ownToken) else {
                        // Registering succeeded on the relay but the secret could not be stored.
                        // Publishing a capability we cannot reproduce would be worse than publishing
                        // none, so this reports failure and the URL goes out without one.
                        completion(nil)
                        return
                    }
                    capQueue.sync { lastAttempt[publicKey] = nil }
                    completion(cap)
                }
            }
            capQueue.sync { pendingChallenges[cid] = { nonce in finish(nonce) } }
            DispatchQueue.global().asyncAfter(deadline: .now() + 30) { finish(nil) }
        }
    }

    /// Called from AppDelegate when the relay's data-only challenge push arrives.
    ///
    /// A challenge nobody asked for is ignored rather than stored: the only party who can cause one
    /// is somebody who knows this token, and there is nothing here for them to gain.
    static func handleRegistrationChallenge(userInfo: [AnyHashable: Any]) -> Bool {
        guard let nonce = userInfo["khandaq_reg_nonce"] as? String,
              let cid = userInfo["khandaq_reg_cid"] as? String,
              !nonce.isEmpty, !cid.isEmpty else {
            return false
        }
        var waiter: ((String) -> Void)?
        capQueue.sync { waiter = pendingChallenges[cid] }
        waiter?(nonce)
        return true
    }

    /// Stop honouring one contact's capability. Presenting it is the proof of the right to drop it.
    static func revokeCapability(forFriend publicKey: String, ownToken: String) {
        let key = capabilityKey(forFriend: publicKey)
        guard let cap = capabilityValue(key), !cap.isEmpty else { return }
        postJSON(relayBase + "/register/revoke", ["token": ownToken, "cap": cap]) { _ in }
        keychainDelete(key)
        // Also clear any pre-KQ-08 copy, so a revoked capability cannot come back from a preference
        // that a migration had not reached yet.
        UserDefaults.standard.removeObject(forKey: key)
    }

    private static func postJSON(_ url: String, _ body: [String: String],
                                 completion: @escaping ([String: Any]?) -> Void) {
        guard let u = URL(string: url),
              let data = try? JSONSerialization.data(withJSONObject: body) else {
            completion(nil)
            return
        }
        var request = URLRequest(url: u)
        request.httpMethod = "POST"
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.httpBody = data
        request.timeoutInterval = 10
        URLSession.shared.dataTask(with: request) { data, response, _ in
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                completion(nil)
                return
            }
            let parsed = data.flatMap { try? JSONSerialization.jsonObject(with: $0) } as? [String: Any]
            completion(parsed ?? [:])
        }.resume()
    }

    private static func hmacSha256Hex(_ secret: String, _ message: String) -> String {
        guard let keyData = secret.data(using: .utf8) else { return "" }
        let msgData = Data(message.utf8)
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        keyData.withUnsafeBytes { keyBytes in
            msgData.withUnsafeBytes { msgBytes in
                CCHmac(
                    CCHmacAlgorithm(kCCHmacAlgSHA256),
                    keyBytes.baseAddress, keyData.count,
                    msgBytes.baseAddress, msgData.count,
                    &digest
                )
            }
        }
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}

File: khandaq-android-trifa/android-refimpl-app/app/src/main/java/org/khandaq/messenger/KhandaqPush.java
package org.khandaq.messenger;

import org.khandaq.messenger.BuildConfig;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/** Khandaq 0.2.0 push relay constants (privacy-preserving wake URLs). */
public final class KhandaqPush {
    private KhandaqPush() {}

    public static final String RELAY_BASE = "https://push.khandaq.org";
    public static final String FCM_PUSH_URL_PREFIX = RELAY_BASE + "/toxfcm/fcm.php?id=";
    public static final String TOKEN_CHANGED_ACTION = "org.khandaq.messenger.TOKEN_CHANGED";
    /** Heads-up channel for FCM wake notifications (created in MainApplication). */
    public static final String FCM_WAKE_CHANNEL_ID = "khandaq_fcm_wake";
    /** Full-screen incoming audio/video call alerts. */
    public static final String INCOMING_CALL_CHANNEL_ID = "khandaq_incoming_call";

    /** Append sender pubkey and a replay-resistant relay auth to a push wake URL. */
    public static String withWakeParams(String pushUrl, String senderPubkeyHex)
    {
        if (pushUrl == null || pushUrl.isEmpty())
        {
            return pushUrl;
        }

        final StringBuilder sb = new StringBuilder(pushUrl);

        if (senderPubkeyHex != null && !senderPubkeyHex.isEmpty() && !pushUrl.contains("from="))
        {
            sb.append(pushUrl.contains("?") ? '&' : '?');
            sb.append("from=").append(senderPubkeyHex);
        }

        // KHANDAQ (security NEW-2): request-bound, time-limited auth instead of a constant HMAC.
        // The sender signs the recipient token (id) + sender (from) + a unix timestamp (ts), so the
        // value differs per request and expires — it cannot be replayed or reused for another push.
        // Dormant until a secret is provisioned (BuildConfig.PUSH_RELAY_AUTH_SECRET empty = no-op).
        final String url = sb.toString();
        final String secret = BuildConfig.PUSH_RELAY_AUTH_SECRET;
        if (secret != null && !secret.isEmpty() && url.contains("push.khandaq.org") && !url.contains("auth="))
        {
            final String id = queryValue(url, "id");
            final String from = queryValue(url, "from");
            final String ts = Long.toString(System.currentTimeMillis() / 1000L);
            // must match the server byte-for-byte: msg = id + "\n" + from + "\n" + ts
            final String auth = hmacSha256Hex(secret, id + "\n" + from + "\n" + ts);
            if (!auth.isEmpty())
            {
                sb.append('&').append("ts=").append(ts).append("&auth=").append(auth);
            }
        }

        return sb.toString();
    }

    /**
     * How to emit one wake, decided from the recipient's push URL.
     *
     * KHANDAQ (re-audit 2026-08-22, K-03). The FCM registration token is a targeting secret — hold
     * it and you can push to that device — and it used to travel in the request URI, next to the
     * HMAC and its timestamp. nginx redacts the query string, but a URL leaks through more than one
     * log: client diagnostics, crash reporters, an intermediary proxy, a copied link, a screenshot.
     * A request BODY leaks through none of those by default.
     *
     * So for our own relay the token, the sender and the capability move into a JSON body and the
     * authentication into headers. Anything else — the legacy tox.zoff.xyz relay, a URL we do not
     * recognise — keeps the old form exactly, because a push URL we do not own is not ours to
     * reinterpret.
     */
    public static final class WakeRequest
    {
        /** Absolute URL to call. */
        public final String url;
        /** JSON body, or null when this is the legacy form-post shape. */
        public final String jsonBody;
        /** Bearer value for Authorization, or null when no secret is provisioned. */
        public final String auth;
        /** Value for X-Khandaq-Ts, or null. */
        public final String ts;

        WakeRequest(String url, String jsonBody, String auth, String ts)
        {
            this.url = url;
            this.jsonBody = jsonBody;
            this.auth = auth;
            this.ts = ts;
        }

        public boolean isJson()
        {
            return jsonBody != null;
        }
    }

    /** True when this URL is our own relay's wake endpoint, and therefore ours to modernise. */
    public static boolean isOwnRelayWakeUrl(final String pushUrl)
    {
        return pushUrl != null
               && pushUrl.startsWith(RELAY_BASE + "/toxfcm/fcm.php")
               && !queryValue(pushUrl, "id").isEmpty();
    }

    public static WakeRequest buildWakeRequest(final String pushUrl, final String senderPubkeyHex)
    {
        if (!isOwnRelayWakeUrl(pushUrl))
        {
            // Unchanged behaviour for every URL that is not ours.
            return new WakeRequest(withWakeParams(pushUrl, senderPubkeyHex), null, null, null);
        }

        final String id = queryValue(pushUrl, "id");
        final String from = (senderPubkeyHex == null) ? "" : senderPubkeyHex;
        // The capability the RECIPIENT minted for us and published inside its wake URL. We carry it
        // across verbatim; we never mint or guess one for somebody else's device.
        final String cap = queryValue(pushUrl, "cap");

        final StringBuilder body = new StringBuilder(128);
        body.append("{\"token\":").append(jsonString(id));
        body.append(",\"sender\":").append(jsonString(from));
        if (!cap.isEmpty())
        {
            body.append(",\"cap\":").append(jsonString(cap));
        }
        body.append('}');

        String auth = null;
        String ts = null;
        final String secret = BuildConfig.PUSH_RELAY_AUTH_SECRET;
        if (secret != null && !secret.isEmpty())
        {
            ts = Long.toString(System.currentTimeMillis() / 1000L);
            // Byte-for-byte the pre-image the legacy path signs: id + "\n" + from + "\n" + ts.
            // Moving the request shape must not move the signature, or a client that switches
            // endpoints looks to the relay like a client with the wrong secret.
            final String mac = hmacSha256Hex(secret, id + "\n" + from + "\n" + ts);
            if (!mac.isEmpty())
            {
                auth = "Bearer " + mac;
            }
            else
            {
                ts = null;
            }
        }
        return new WakeRequest(RELAY_BASE + "/wake", body.toString(), auth, ts);
    }

    /** Minimal JSON string escaping. The values here are base64url/hex, but assuming that is how a
     *  quote ends up unescaped in a body the relay then fails to parse. */
    public static String jsonString(final String s)
    {
        final StringBuilder out = new StringBuilder(s.length() + 2);
        out.append('"');
        for (int i = 0; i < s.length(); i++)
        {
            final char c = s.charAt(i);
            switch (c)
            {
                case '"':  out.append("\\\""); break;
                case '\\': out.append("\\\\"); break;
                case '\n': out.append("\\n"); break;
                case '\r': out.append("\\r"); break;
                case '\t': out.append("\\t"); break;
                default:
                    if (c < 0x20)
                    {
                        out.append(String.format("\\u%04x", (int) c));
                    }
                    else
                    {
                        out.append(c);
                    }
            }
        }
        out.append('"');
        return out.toString();
    }

    /** Extract the URL-decoded value of a query parameter (to match the server's request.args view). */
    private static String queryValue(final String url, final String key)
    {
        try
        {
            final int q = url.indexOf('?');
            if (q < 0)
            {
                return "";
            }
            for (final String pair : url.substring(q + 1).split("&"))
            {
                final int eq = pair.indexOf('=');
                if (eq > 0 && pair.substring(0, eq).equals(key))
                {
                    return java.net.URLDecoder.decode(pair.substring(eq + 1), "UTF-8");
                }
            }
        }
        catch (Exception ignored)
        {
        }
        return "";
    }

    private static String hmacSha256Hex(final String secret, final String message)
    {
        try
        {
            final Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secret.getBytes(java.nio.charset.StandardCharsets.UTF_8), "HmacSHA256"));
            final byte[] raw = mac.doFinal(message.getBytes(java.nio.charset.StandardCharsets.UTF_8));
            final StringBuilder hex = new StringBuilder(raw.length * 2);
            for (byte b : raw)
            {
                hex.append(String.format("%02x", b));
            }
            return hex.toString();
        }
        catch (Exception ignored)
        {
            return "";
        }
    }
}

